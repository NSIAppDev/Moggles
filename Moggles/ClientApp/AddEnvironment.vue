<template>
    <div>
        <spinner ref="spinner" v-model="spinner" size="sm"></spinner>

        <alert v-model="showSuccessAlert" placement="top-right" duration="1500" type="success" width="400px" dismissable>
            <span class="icon-ok-circled alert-icon-float-left"></span>
            <p>Environment added.</p>
        </alert>

        <div class="panel-body">
            <div class="">
                <div class="form-group">
                    <div v-for="error in errors" :key="error" class="validationMessage">{{error}}</div>
                    <input class="form-control" v-model="envName" type="text" name="envName" placeholder="Env name..." maxlength="50">
                </div>
                <div class="form-group">
                    <div>
                        Default toggle value:
                    </div>
                    <div>
                        <label for="d1">True</label>
                        <input id="d1" v-model="defaultToggleValue" type="radio" name="defToggleValue" value="true" checked>

                        <label for="d2">False</label>
                        <input id="d2" v-model="defaultToggleValue" type="radio" name="defToggleValue" value="false">
                    </div>
                </div>
                <button :disabled="applicationId > 0? false : true" class="btn btn-default btn-primary" v-on:click="addEnv" type="button">Add</button>
            </div>
        </div>
    </div>
</template>

<script>
    import { Bus } from './event-bus'
    import axios from 'axios'
    import { spinner, alert } from 'vue-strap'

    export default {
        data() {
            return {
                applicationId: -1,
                envName: "",
                sortOrder: 500,
                defaultToggleValue: true,
                existingEnvs: [],
                errors: [],
                spinner: false,
                showSuccessAlert: false
            }
        },
        methods: {
            addEnv() {
                if (this.applicationId === -1)
                    return;

                this.errors = [];

                if (this.existingEnvs.some(env => env === this.envName)) {
                    this.errors.push("Environment already exists")
                    return;
                }

                if (this.envName === "") {
                    this.errors.push("Environment name cannot be empty")
                    return;
                }

                let param = {
                    applicationId: this.applicationId,
                    envName: this.envName,
                    sortOrder: this.sortOrder,
                    defaultToggleValue: this.defaultToggleValue
                }

                this.spinner = true;
                axios.post('api/FeatureToggles/AddEnvironment', param)
                    .then((response) => {
                        this.showSuccessAlert = true;
                        this.envName = ''
                        Bus.$emit("env-added")
                    }).catch((e) => {
                        window.alert(e)
                    }).finally(() => {
                        this.spinner = false;
                    });
            }
        },
        mounted() {
            Bus.$on("app-changed", app => {
                this.applicationId = app.id;
            });

            Bus.$on("env-loaded", envs => {
                this.existingEnvs = envs;
            });
        },
        components: {
            spinner,
            alert
        }
    }
</script>

<style>

    .validationMessage {
        color: red
    }
</style>