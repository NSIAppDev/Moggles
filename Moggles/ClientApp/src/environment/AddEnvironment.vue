<template>
  <div>
    <alert v-if="showSuccessAlert" :duration="alertDuration" type="success"
           @dismissed="showSuccessAlert = false">
      <p>
        <i class="fas fa-check-circle" /> Environment added.
      </p>
    </alert>

    <div class="form-horizontal">
      <div class="row">
        <div v-for="error in errors" :key="error" class="text-danger">
          {{ error }}
        </div>
        <div class="col-sm-12 form-group">
          <label class="col-sm-4 control-label" for="envname">Environment name</label>
          <div class="col-sm-7">
            <input ref="envName" v-model="envName" class="col-sm-8 form-control"
                   type="text"
                   name="envName" placeholder="Env name..." maxlength="50">
          </div>
        </div>
        <div class="col-sm-12 form-group">
          <label class="col-sm-4 control-label">
            Default toggle value
          </label>
          <div class="col-sm-6 margin-top-4">
            <label for="b1">True</label>
            <input id="b1" v-model="defaultToggleValue" type="radio"
                   :value="true" checked>

            <label for="b2">False</label>
            <input id="b2" v-model="defaultToggleValue" type="radio"
                   :value="false">
          </div>
        </div>
        <div class="col-sm-12 form-group">
          <label class="col-sm-4 control-label">
            Require a reason when toggle state changes to 
          </label>
          <div class="col-sm-6 margin-top-4">
            <label for="requireReasonWhenToggleEnabled">Enabled</label>
            <p-check v-model="requireReasonWhenToggleEnabled" class="p-icon p-fill" color="default">
              <i slot="extra" class="icon fas fa-check" />
            </p-check>
            <label for="requireReasonWhenToggleDisabled">Disabled</label>
            <p-check v-model="requireReasonWhenToggleDisabled" class="p-icon p-fill" color="default">
              <i slot="extra" class="icon fas fa-check" />
            </p-check>
          </div>
        </div>
        <div class="col-sm-12 text-right">
          <button class="btn btn-default" @click="closeAddEnvironmentModal">
            Close
          </button>
          <button :disabled="application.id != ''? false : true" class="btn btn-primary" type="button"
                  @click="addEnv">
            Add
          </button>
        </div>
      </div>
    </div>
  </div>
</template>

<script>
    import { Bus } from '../common/event-bus'
    import axios from 'axios'
    import PrettyCheck from 'pretty-checkbox-vue/check'

    export default {
        components: {
            'p-check': PrettyCheck
        },
        props: {
            application: {
                type: Object,
                required: true
            }
        },
        data() {
            return {
                envName: "",
                sortOrder: 500,
                defaultToggleValue: false,
                existingEnvs: [],
                errors: [],
                showSuccessAlert: false,
                alertDuration: 1500,
                requireReasonWhenToggleEnabled: false,
                requireReasonWhenToggleDisabled:false
            }
        },
        mounted() {
            Bus.$on("env-loaded", envs => {
                this.existingEnvs = envs;
            });

            this.$nextTick(() => { this.$refs["envName"].focus() });
            this.clearFields();
        },
        methods: {
            clearFields() {
                this.envName = "";
                this.errors = [];
                this.defaultToggleValue = false;
            },
            addEnv() {
                if (this.application.id === -1)
                    return;

                this.errors = [];

                if (this.existingEnvs.some(env => env === this.envName)) {
                    this.errors.push("Environment already exists")
                    return;
                }

                if (this.envName === "") {
                    this.errors.push("Environment name cannot be empty")
                    return;
                }
                
                let param = {
                    applicationId: this.application.id,
                    envName: this.envName,
                    sortOrder: this.sortOrder,
                    defaultToggleValue: this.defaultToggleValue,
                    requireReasonToChangeWhenToggleEnabled: this.requireReasonWhenToggleEnabled,
                    requireReasonToChangeWhenToggleDisabled: this.requireReasonWhenToggleDisabled
                }

                Bus.$emit('block-ui')
                axios.post('api/FeatureToggles/AddEnvironment', param)
                    .then(() => {
                        this.showSuccessAlert = true;
                        this.envName = '';
                        this.defaultToggleValue = false;
                        this.requireReasonWhenToggleEnabled = false;
                        this.requireReasonWhenToggleDisabled = false;
                        this.$nextTick(() => { this.$refs["envName"].focus() });
                        Bus.$emit("env-added")
                    }).catch((e) => {
                        window.alert(e)
                    }).finally(() => {
                        Bus.$emit('unblock-ui')
                    });
            },
            closeAddEnvironmentModal() {
                Bus.$emit('close-add-environment');
            }
        }
    }
</script>