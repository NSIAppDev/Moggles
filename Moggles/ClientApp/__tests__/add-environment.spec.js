import { mount } from '@vue/test-utils'
import AddEnvironment from '../src/environment/AddEnvironment.vue'


describe('AddEnvironment.vue', () => {
    test('is a Vue instance', () => {
        const wrapper = mount(AddEnvironment, {
            propsData: {
                application: {
                    id:1
                }
            }
        })
        expect(wrapper.findComponent(AddEnvironment).vm).toBeTruthy()
    })
})