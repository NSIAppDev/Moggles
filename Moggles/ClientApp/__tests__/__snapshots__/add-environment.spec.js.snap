// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddEnvironment.vue matches snapshot 1`] = `
<div>
  <div
    class="spinner spinner-gritcode spinner-sm"
    style="display: none;"
  >
    <div
      class="spinner-wrapper"
    >
      <div
        class="spinner-circle"
      />
       
      <div
        class="spinner-text"
      >
        
      </div>
    </div>
  </div>
   
  <div
    class="alert alert-success top-right"
    name="fade"
    role="alert"
    style="width: 400px; display: none;"
  >
    <button
      class="close"
      type="button"
    >
      <span>
        Ã—
      </span>
    </button>
     
    <span
      class="icon-ok-circled alert-icon-float-left"
    />
     
    <p>
      Environment added.
    </p>
  </div>
   
  <div
    class="panel-body"
  >
    <div>
      <div
        class="form-group"
      >
         
        <input
          class="form-control"
          maxlength="50"
          name="envName"
          placeholder="Env name..."
          type="text"
        />
      </div>
       
      <div
        class="form-group"
      >
        <div>
          
                    Default toggle value:
                
        </div>
         
        <div>
          <label
            for="d1"
          >
            True
          </label>
           
          <input
            checked="checked"
            id="d1"
            name="defToggleValue"
            type="radio"
            value="true"
          />
           
          <label
            for="d2"
          >
            False
          </label>
           
          <input
            id="d2"
            name="defToggleValue"
            type="radio"
            value="false"
          />
        </div>
      </div>
       
      <button
        class="btn btn-default btn-primary"
        disabled="disabled"
        type="button"
      >
        Add
      </button>
    </div>
  </div>
</div>
`;
